src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -0,0 +1,124 @@
+
+            return new VisualStudioInstance(
+                name: ".NET Core SDK",
+                path: dotNetSdkPath,
+                version: new Version(fileVersionInfo.FileMajorPart, fileVersionInfo.FileMinorPart, fileVersionInfo.FileBuildPart),
+                discoveryType: DiscoveryType.DotNetSdk);
+        }
+
+        private static string hostfxr_resolve_sdk(string workingDir)
+        {
+            Process process = Process.GetCurrentProcess();
+
+            string dotNetExePath = process?.MainModule?.FileName;
+
+            if (String.IsNullOrWhiteSpace(dotNetExePath))
+            {
+                return null;
+            }
+
+            string dotNetExeDirectory = Path.GetDirectoryName(dotNetExePath);
+
+            if (String.IsNullOrWhiteSpace(dotNetExeDirectory) || !Directory.Exists(dotNetExeDirectory))
+            {
+                return null;
+            }
+
+            var buffer = new StringBuilder(capacity: 64);
+
+            for (;;)
+            {
+                int size = Interop.hostfxr_resolve_sdk(dotNetExeDirectory, workingDir, buffer, buffer.Capacity);
+                if (size <= 0)
+                {
+                    return null;
+                }
+
+                if (size <= buffer.Capacity)
+                {
+                    break;
+                }
+
+                buffer.Capacity = size;
+            }
+
+            return buffer.ToString();
+        }
+
+        private static class Interop
+        {
+            private static readonly bool RunningOnWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+
+            public static string realpath(string path)
+            {
+                var ptr = unix_realpath(path, IntPtr.Zero);
+                var result = Marshal.PtrToStringAnsi(ptr); // uses UTF8 on Unix
+                unix_free(ptr);
+                return result;
+            }
+
+            public static int hostfxr_resolve_sdk(string exeDir, string workingDir, [Out] StringBuilder buffer, int bufferSize)
+            {
+                // hostfxr string encoding is platform -specific so dispatch to the
+                // appropriately annotated P/Invoke for the current platform.
+                return RunningOnWindows
+                    ? windows_hostfxr_resolve_sdk(exeDir, workingDir, buffer, bufferSize)
+                    : unix_hostfxr_resolve_sdk(exeDir, workingDir, buffer, bufferSize);
+            }
+
+            [DllImport("hostfxr", EntryPoint = nameof(hostfxr_resolve_sdk), CharSet = CharSet.Unicode, ExactSpelling = true, CallingConvention = CallingConvention.Cdecl)]
+            private static extern int windows_hostfxr_resolve_sdk(string exeDir, string workingDir, [Out] StringBuilder buffer, int bufferSize);
+
+            // CharSet.Ansi is UTF8 on Unix
+            [DllImport("hostfxr", EntryPoint = nameof(hostfxr_resolve_sdk), CharSet = CharSet.Ansi, ExactSpelling = true, CallingConvention = CallingConvention.Cdecl)]
```

Comment:
Did you try this on Unix yet? I don't believe the P/invoke will just bind to the hostfxr that's already been loaded on non-Windows.

Reply:
I see above that you had it working on Ubuntu. I am confused now. Maybe it was only on OS X where I saw this behavior, See https://github.com/dotnet/core-setup/pull/4301/commits/d438d9838afcb7a15e86348ff1a2f8084e54e206 where I had to copy libhostfxr for tests.

Reply:
Thinking more, I'm 99% sure I've seen this behavior difference on Linux too. I'm very confused.

Reply:
I can repro it working on Ubuntu and failing on Mac OS X. From the discussion [here](https://github.com/dotnet/coreclr/issues/12707), I would not expect it to work on Ubuntu either, but it does.